File: python/crawlab/db/sql.py
import psycopg2
import pymysql

from crawlab.constants import DataSourceType
from crawlab.utils.config import get_data_source, get_data_source_type, get_collection

conn_cache = {}


def get_conn():
    ds_type = get_data_source_type()
    if conn_cache.get(ds_type) is None:
        if ds_type == DataSourceType.MYSQL:
            # MySQL
            conn_cache[ds_type] = connect_mysql()
        elif ds_type == DataSourceType.POSTGRES:
            # Postgres
            conn_cache[ds_type] = connect_postgres()
    return conn_cache.get(ds_type)


def connect_mysql():
    ds = get_data_source()
    return pymysql.connect(
        host=ds.get('host') or 'localhost',
        port=int(ds.get('port')) if ds.get('port') is not None else 3306,
        user=ds.get('username') or None,
        password=ds.get('password') or None,
        database=ds.get('database') or 'test',
        charset='utf8'
    )


def connect_postgres():
    ds = get_data_source()
    return psycopg2.connect(
        host=ds.get('host') or 'localhost',
        port=ds.get('port') or '5432',
        user=ds.get('username') or None,
        password=ds.get('password') or None,
        database=ds.get('database') or 'test',
    )


def _insert_item(item: dict):
    conn = get_conn()
    columns = item.keys()
    table_name = get_collection()
    if table_name is None:
        raise Exception('table_name is empty')
    sql_str = f'INSERT INTO {table_name}({",".join(columns)}) VALUES ({",".join(["%s" for _ in columns])});'
    cursor = conn.cursor()
    cursor.execute(sql_str, list(item.values()))
    conn.commit()
    cursor.close()


def insert_item_mysql(item: dict):
    _insert_item(item)


def insert_item_postgres(item: dict):
    _insert_item(item)


def insert_item_sqlserver(item: dict):
    raise NotImplementedError('sqlserver is not implemented')


def insert_item(item: dict):
    ds_type = get_data_source_type()
    if ds_type == DataSourceType.MYSQL:
        insert_item_mysql(item)
    elif ds_type == DataSourceType.POSTGRES:
        insert_item_postgres(item)
    elif ds_type == DataSourceType.SQLSERVER:
        insert_item_sqlserver(item)
    else:
        raise NotImplementedError(f'{ds_type} is not implemented')

File: go/database/sql.go
package database

import (
	"errors"
	"fmt"
	"github.com/apex/log"
	"github.com/crawlab-team/crawlab-sdk/go/constants"
	"github.com/crawlab-team/crawlab-sdk/go/entity"
	"github.com/crawlab-team/crawlab-sdk/go/utils"
	"github.com/jmoiron/sqlx"
	"runtime/debug"
	"strings"
)

var Db *sqlx.DB

func GetSqlDatabaseConnectionString(dataSourceType string, host string, port string, username string, password string, database string) (connStr string, err error) {
	// 获取数据库链接串
	if dataSourceType == constants.DataSourceTypeMysql {
		connStr = fmt.Sprintf("%s:%s@(%s:%s)/%s?charset=utf8&parseTime=True&loc=Local", username, password, host, port, database)
	} else if dataSourceType == constants.DataSourceTypePostgres {
		connStr = fmt.Sprintf("host=%s port=%s user=%s dbname=%s password=%s sslmode=%s", host, port, username, database, password, "disable")
	} else {
		err = errors.New(dataSourceType + " is not implemented")
		log.Errorf(err.Error())
		debug.PrintStack()
		return connStr, err
	}
	return connStr, nil
}

func GetSqlDatabase() *sqlx.DB {
	if Db != nil {
		return Db
	}

	ds := utils.GetDataSource()

	// 获取数据库链接串
	connStr, err := GetSqlDatabaseConnectionString(ds.Type, ds.Host, ds.Port, ds.Username, ds.Password, ds.Database)
	if err != nil {
		log.Errorf("get connection string error: " + err.Error())
		debug.PrintStack()
		return Db
	}

	// 数据库
	Db, err = sqlx.Open(ds.Type, connStr)
	if err != nil {
		log.Errorf("open database error: " + err.Error())
		debug.PrintStack()
	}

	return Db
}

func InsertItem(item entity.Item) (err error) {
	db := GetSqlDatabase()
	col := utils.GetCollection()
	if _, err := db.Exec(fmt.Sprintf("INSERT INTO %s (%s) VALUES (%s)",
		col,
		strings.Join(utils.GetItemKeys(item), ","),
		strings.Join(utils.GetItemValuesWithQuote(item), ","),
	)); err != nil {
		log.Errorf("insert item error: " + err.Error())
		debug.PrintStack()
		return err
	}
	return nil
}

func UpdateItem(item entity.Item, dedupField string) (err error) {
	db := GetSqlDatabase()
	col := utils.GetCollection()
	if _, err := db.Exec(fmt.Sprintf("UPDATE %s SET %s WHERE %s = '%s'",
		col,
		utils.GetItemSqlUpdateStr(item),
		dedupField,
		item[dedupField],
	)); err != nil {
		log.Errorf("update item error: " + err.Error())
		debug.PrintStack()
		return err
	}
	return nil
}

func GetItem(key string, value string) (item entity.Item, err error) {
	db := GetSqlDatabase()
	col := utils.GetCollection()
	if err := db.QueryRow(fmt.Sprintf("SELECT * FROM %s WHERE %s = '%s'",
		col,
		key,
		value,
	)).Scan(&item); err != nil {
		log.Errorf("update item error: " + err.Error())
		debug.PrintStack()
		return item, err
	}
	return item, nil
}
